contracts/stake/src/state.rs

```diff
@@ -56,28 +56,26 @@ impl StakeState {
 
     pub fn stake(&mut self, stake: Stake) {
         let value = stake.value();
-        let keys = *stake.keys();
-        let account = &keys.account;
+        let mut keys = *stake.keys();
         let signature = *stake.signature();
 
         if stake.chain_id() != self.chain_id() {
             panic!("The stake must target the correct chain");
         }
 
-        let prev_stake = self.get_stake(account).copied();
+        let prev_stake = self.get_stake(&keys.account).copied();
         let (loaded_stake, loaded_keys) = self.load_or_create_stake_mut(&keys);
 
-        if loaded_stake.amount.is_some() {
-            panic!("Can't stake twice for the same key");
+        if loaded_stake.is_empty() {
+            // Update the funds key with the newly provided one
+            // This operation will rollback if the signature is invalid
+            *loaded_keys = keys;
+        } else {
+            // Use the existing keys to verify signatures
+            keys = *loaded_keys;
```

Comment:
I'm not sure about this overriding of keys.

After calling
```rust
let (loaded_stake, loaded_keys) = self.load_or_create_stake_mut(&keys);
```
we know for sure that `keys.account == loaded_keys.account`, since the stakes are stored with the `account` key as map-key.

This means that the in the `if` statement we would only override the `funds` keys should they be different.

In the first part
```rust
        if loaded_stake.is_empty() {
            // Update the funds key with the newly provided one
            // This operation will rollback if the signature is invalid
            *loaded_keys = keys;
        }
```
We would override the `funds` key that is stored on-chain with the given `funds` key from the stake-operation.
Given the fact that we remove empty stakes, this seems superfluous as an empty stake suggests that the stake has just been created in `load_or_create_stake_mut`, which means that the keys are the same anyway.
In the unlikely but maybe not impossible case that the stake is empty and `keys.funds != loaded_keys.funds`, the overriding of the stored keys is a behavior that doesn't seem to do any harm but that I think should at least be clearly documented.

Coming to the second part of the `if` statement:
```rust
        else {
            // Use the existing keys to verify signatures
            keys = *loaded_keys;
        }
```
This overriding makes sure that the funds signature will be verified by the funds key that is stored with the stake.
This is desirable behavior but also a bit hidden logic.


Long story short, I propose something in the lines of the following instead of the `if`-statement:
```suggestion
if loaded_keys.funds != keys.funds {
    // if the funds keys are different and the stake is empty we update the stored funds key with the provided one
    if loaded_stake.is_empty() {
        *loaded_keys = keys;
    } else {
        panic!("The provided funds keys doesn't match the funds key stored on chain");
    }
}
```
This would simplify the logic of this function and give a more descriptive panic message if the keys don't match.

Reply:
This is a leftover implementation that was needed when the stake entry was never removed (when we have the nonce)

I'm going to change this, to let the load_or_create_stake_mut to be in charge of return the proper keys.

However, I'm not sure to want to check that the keys must be the same, this check is already done implicitly by the signature verification

Reply:
I've moved the check in the load_or_create_stake_mut